### PR TITLE
Optimize bindings

### LINKED ISSUE #9: Optimize bindings
Optimize and refactor the existing bindings:
* Optimize the build process
* Make use of base classes for `Block` and `File`
* Throw nicer exceptions etc
* Implement `equals()`, `hasCode()` and `compareTo()`

### PR DESCRIPTION
**Pull request by sujithvm**
@stoewer fix #9 Can you review the overrides for Block? Thank you

### CODE REVIEW

**Comment by stoewer** [comment]
The overrides look fine for me.

#9 also mentions bindings for super classes for `File` and `Block` such as `ImplContainer`, `Entity`, `NamedEntity` etc., have you already looked into this?

**Comment by sujithvm** [comment]
@stoewer Should the base classes also be wrapped? Currently I have implemented the base methods in Block.

**Comment by stoewer** [comment]
Since allot of NIX classes inherit from `ImplContainer`, `Entity`, `NamedEntity` etc. bindings for this super classes would reduce code duplication quite a bit. If it can be done with JavaCPP (I was not able to google any examples) I would strongly recommend it.

**Comment by sujithvm** [comment]
@stoewer It can be done by making a separate class for each. For example OptionalString was made to wrap boost::optional<std::string> . I shall check if there is any other alternative way of doing so.

**Comment by stoewer** [comment]
If it can only be done via a binding for each specialization, it is not worth doing it. But it is worth finding out if this is actually the case.

**Comment by sujithvm** [comment]
@stoewer What changes need to be done with respect to optimization and exceptions? Thank you

**Comment by stoewer** [comment]
@sujithvm as I mentioned above: the inheritance problem has to be solved somehow. Either via interfaces that reflect the bound methods of the respective super class or (if possible) by bindings to the NIX base classes.

**Comment by stoewer** [comment]
@sujithvm I found another problem related to this pull request which we have to address. Most NIX entities have a method `isNone()`. This method is necessary because we handle NIX objects often by value and is an equivalent for `nullptr` or `NULL` in c++. However, in java everything is a reference and there are two ways to represent null values in Java: one is `null` and the other is java-8 `Optional<T>`.

If you look at the `getBlock(String)` method you will see that it always returns a Block and never throws exceptions. Instead if a Block with a name or id exists `isNone()` (or `! isInitialized()`) is `false` but if the block does not exists it is `true`.

The question is now: is this representation obvious for a Java developer or would he expect something like `null` or an empty `Optional`. If the latter is the case methods like `getBlock()` should be wrapped and convert the Block to the desired null representation if it is not initialized.

**Comment by sujithvm** [comment]
@stoewer I just faced that issue while working on an other issue. While working on test for Block metadata access https://github.com/sujithvm/nix-java/blob/issue14/src/test/java/org/gnode/nix/TestBlock.java  I encountered this issue. So I changed the `getMetadata()` function as seen here https://github.com/sujithvm/nix-java/blob/issue14/src/main/java/org/gnode/nix/Block.java

**Comment by sujithvm** [comment]
@stoewer Should I change the functions to return a null? As optional<T> is available only in Java 8.

**Comment by stoewer** [comment]
`Optional<T>` being available only for Java 8 is not really an issue, if it is beneficial for some reason, I would not hesitate restricting the bindings to Java 8.

The more important question is: what is nicer to use? Using `Optional<T>` has many advantages, but if you start using it in one place of your public API you should use it everywhere (if something can be `null`) e.g. the setters that bind to `removeSomething()`. And I am not sure whether I like that.

Just for now I would go with the `null` option but maybe we can revisit this topic later (I'll create an issue).

**Comment by sujithvm** [comment]
@stoewer Okay. I shall investigate on the usage of optional<T> over null.

**Comment by stoewer** [comment]
@sujithvm Actually I meant "Just for now I would go with the `null` option ..."

However, I created and issue #27 for M3 so we can think about this again later when we see the full impact on the API.

**Comment by sujithvm** [comment]
@stoewer Regarding bindings for super classes it seems like a binding must be made for each specialization. So I have decided to proceed with Java interfaces but I am facing another issue with using  plain interfaces with JavaCPP which was reported here https://groups.google.com/forum/#!topic/javacpp-project/pXRQ_UcWfs8 . I have also supported the issue. I am waiting to hear from the developer.

**Comment by sujithvm** [comment]
@stoewer Okay. I shall proceed with using `null` for now.

**Comment by sujithvm** [comment]
@stoewer It seems like Java interface issue has been fixed by the developer in a commit. But this wont available until next release. So should I proceed with building JavaCPP from source and add as an external dependency like before?

**Comment by stoewer** [comment]
@sujithvm Or we use abstract classes with only abstract methods instead of interfaces for now and refactor them into interfaces as soon as the new release is out. What would you prefer? Is there any comment from the devs about the time frame for the next release?

**Comment by sujithvm** [comment]
@stoewer Okay. I shall proceed with using abstract methods till next release is out. I shall also check when the next release is expected to be out.

**Comment by sujithvm** [comment]
@stoewer Can you review the base abstract classes ? Thank you.

File: src/main/java/org/gnode/nix/base/Entity.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    abstract public void setCreatedAt();
+
+    /**
+     * Sets the creation time to the provided value even if the
+     * field is already set.
+     *
+     * @param date The creation time to set.
+     */
+    abstract public void forceCreatedAt(Date date);
```
Comment (stoewer) [inline src/main/java/org/gnode/nix/base/Entity.java]: You can avoid some code duplication if you implement `equals()` and `hasCode()` here.

File: src/main/java/org/gnode/nix/base/NamedEntity.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * Getter for the definition of the entity.
+     * <p/>
+     * The definition is an optional property that allows the user to add
+     * a freely assignable textual definition to the entity.
+     *
+     * @return The definition of the entity.
+     */
+    abstract public String getDefinition();
```
Comment (stoewer) [inline src/main/java/org/gnode/nix/base/NamedEntity.java]: You can implement `compareTo()` here.

File: src/main/java/org/gnode/nix/base/ImplContainer.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+     * @return true if initialized else false
+     */
+    abstract public boolean isNone();
+
+    /**
+     * Checks if entity is initialized
+     *
+     * @return true if initialized else false
+     */
+    abstract public boolean isInitialized();
```
Comment (stoewer) [inline src/main/java/org/gnode/nix/base/ImplContainer.java]: You can implement this base on `isNone()`.

**Comment by stoewer** [comment]
Since we have abstract classes and not interfaces we might as well make use of it and implement some methods in order to avoid code duplication.

**Comment by sujithvm** [comment]
@stoewer  These functions are/use native functions and they can be accessed via bindings from their child class. For eg File -> isNone or Block -> isNone. So if isNone is defined in ImplContainer it cannot determine if its called from File or Block as a binding wrapper cannot be made. So how can this be implemented?

 I shall also check with the JavaCPP devs if there is an elegant solution to the problem.

**Comment by stoewer** [comment]
@sujithvm What I meant was exactly what you did in 05d6a87d6ff2eda41bd6267dcc028658dfa10572: you can implement `isInitialized()` as `! isNone()` but of course `isNone()` itself must be a native method that is bound in each sub class individually.

**Comment by sujithvm** [comment]
@stoewer Sorry about the confusion. Can you please review the changes? Thank you

**Comment by sujithvm** [comment]
@stoewer compareTo() is now fixed.

**Pull request merged**